### PR TITLE
feat(dictionaries): dictionaries API implementation

### PR DESCRIPTION
**Pull request by Aallam**
Implement dictionaries API
- [x] `saveDictionaryEntries`
- [x] `replaceDictionaryEntries`
- [x] `deleteDictionaryEntries`
- [x] `clearDictionaryEntries`
- [x] `searchDictionaryEntries`
- [x] `setDictionarySettings`
- [x] `getDictionarySettings`

### CODE REVIEW

**Review by chloelbn — CHANGES_REQUESTED** [review]

**Comment by VladislavFitz** [comment]
I like the way you introduced the type safety for dictionaries methods. I started to do it in the same way in Swift, but when I found a type structure improvement that you may wish to apply to your implementation.

The questionable thing is the `DictionaryEntry<T>` structure. It resolves the type safety challenge, but semantically I think it is more consistent to make `DictionaryEntry` the descendant of a `Dictionary` and not the other way around.
The idea is that `Dictionary` is not just a string wrapper, but an interface providing a static `name` string and a generic  `DictionaryEntry` parameter.

```kotlin
interface Dictionary<Entry: DictionaryEntry> {
  // not sure how to define a static `name` a variable in the Kotlin interface
}
```

Therefore, each concrete dictionary is an implementation of this protocol:

```kotlin
class CompoundsDictionary: Dictionary<CompoundEntry>
```

As a result, the `Dictionary` parameter explicitly defines the type of its `Entry` and can be used directly in the methods as a `dictionary` parameter.
By doing so, I suppose, you can abolish the type-explicit methods such as `savePluralsEntries`, `deletePluralsEntries`, `searchPluralsEntries` by keeping their generic version.

I applied this approach at the last moment in the Swift implementation and it seems work well. You can check out the [Swift PR](https://github.com/algolia/algoliasearch-client-swift/pull/717/files) (you have to review it anyway 😁) to better understand my idea. 

What do you think?

File: src/commonMain/kotlin/com/algolia/search/endpoint/EndpointDictionary.kt [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+     * @param dictionary target dictionary.
+     * @param dictionaryEntries dictionary entries to be saved.
+     * @param clearExistingDictionaryEntries when `true`, start the batch by removing all the custom entries
+     * from the dictionary.
+     * @param requestOptions Configure request locally with [RequestOptions].
+     */
+    public suspend fun <T : Dictionary> saveDictionaryEntries(
+        dictionary: T,
+        dictionaryEntries: List<DictionaryEntry<T>>,
+        clearExistingDictionaryEntries: Boolean = false,
```
[the remark this answers is not in the document]
Comment (Aallam) [reply src/commonMain/kotlin/com/algolia/search/endpoint/EndpointDictionary.kt:35]: Not sure if I am getting this right, should we the parameter from `saveDictionaryEntries` and set it always to false in the request?

[the remark this answers is not in the document]
Comment (chloelbn) [reply src/commonMain/kotlin/com/algolia/search/endpoint/EndpointDictionary.kt:35]: yes indeed! it'll be true only for replace

[the remark this answers is not in the document]
Comment (Aallam) [reply src/commonMain/kotlin/com/algolia/search/endpoint/EndpointDictionary.kt]: Fixed 🎉

**Review by VladislavFitz — APPROVED** [review]

**Review by chloelbn — APPROVED** [review]

**Review by chloelbn — APPROVED** [review]

**Pull request merged**